# Исследование надежности заемщиков.

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Oписание-проекта" data-toc-modified-id="Oписание-проекта-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Oписание проекта</a></span></li><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Заполнение-пропусков" data-toc-modified-id="Заполнение-пропусков-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Заполнение пропусков</a></span></li><li><span><a href="#Проверка-данных-на-аномалии-и-исправления." data-toc-modified-id="Проверка-данных-на-аномалии-и-исправления.-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Проверка данных на аномалии и исправления.</a></span></li><li><span><a href="#Изменение-типов-данных." data-toc-modified-id="Изменение-типов-данных.-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Изменение типов данных.</a></span></li><li><span><a href="#Удаление-дубликатов." data-toc-modified-id="Удаление-дубликатов.-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Удаление дубликатов.</a></span></li><li><span><a href="#Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма." data-toc-modified-id="Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма.-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.</a></span></li><li><span><a href="#Категоризация-дохода." data-toc-modified-id="Категоризация-дохода.-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Категоризация дохода.</a></span></li><li><span><a href="#Категоризация-целей-кредита." data-toc-modified-id="Категоризация-целей-кредита.-3.7"><span class="toc-item-num">3.7&nbsp;&nbsp;</span>Категоризация целей кредита.</a></span></li></ul></li><li><span><a href="#Исследовательский-анализ.-Ответы-на-вопросы." data-toc-modified-id="Исследовательский-анализ.-Ответы-на-вопросы.-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Исследовательский анализ. Ответы на вопросы.</a></span><ul class="toc-item"><li><span><a href="#Вопрос-1:-Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-1:-Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?</a></span><ul class="toc-item"><li><span><a href="#Вывод-1:" data-toc-modified-id="Вывод-1:-4.1.1"><span class="toc-item-num">4.1.1&nbsp;&nbsp;</span>Вывод 1:</a></span></li></ul></li><li><span><a href="#Вопрос-2:Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-2:Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Вопрос 2:Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span><ul class="toc-item"><li><span><a href="#Вывод-2:" data-toc-modified-id="Вывод-2:-4.2.1"><span class="toc-item-num">4.2.1&nbsp;&nbsp;</span>Вывод 2:</a></span></li></ul></li><li><span><a href="#Вопрос-3:-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-3:-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span><ul class="toc-item"><li><span><a href="#Вывод-3:" data-toc-modified-id="Вывод-3:-4.3.1"><span class="toc-item-num">4.3.1&nbsp;&nbsp;</span>Вывод 3:</a></span></li></ul></li><li><span><a href="#Вопрос-4:-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Вопрос-4:-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Вопрос 4: Как разные цели кредита влияют на его возврат в срок?</a></span><ul class="toc-item"><li><span><a href="#Вывод-4:" data-toc-modified-id="Вывод-4:-4.4.1"><span class="toc-item-num">4.4.1&nbsp;&nbsp;</span>Вывод 4:</a></span></li></ul></li><li><span><a href="#Вопрос-5:-Есть-ли-зависимость-между-уровнем-образования-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-5:-Есть-ли-зависимость-между-уровнем-образования-и-возвратом-кредита-в-срок?-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Вопрос 5: Есть ли зависимость между уровнем образования и возвратом кредита в срок?</a></span><ul class="toc-item"><li><span><a href="#Вывод-5:" data-toc-modified-id="Вывод-5:-4.5.1"><span class="toc-item-num">4.5.1&nbsp;&nbsp;</span>Вывод 5:</a></span></li></ul></li><li><span><a href="#Вопрос-6:-Есть-ли-зависимость-между-типом-занятости-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-6:-Есть-ли-зависимость-между-типом-занятости-и-возвратом-кредита-в-срок?-4.6"><span class="toc-item-num">4.6&nbsp;&nbsp;</span>Вопрос 6: Есть ли зависимость между типом занятости и возвратом кредита в срок?</a></span><ul class="toc-item"><li><span><a href="#Вывод-6:" data-toc-modified-id="Вывод-6:-4.6.1"><span class="toc-item-num">4.6.1&nbsp;&nbsp;</span>Вывод 6:</a></span></li></ul></li></ul></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общий вывод:</a></span></li></ul></div>

## Oписание проекта
**Заказчик** — кредитный отдел банка. 

**Цель исследования** — разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. 

**Входные данные** от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Обзор данных

Ознакомимся с данными. Для этого откроем файл, посмотрим первые 15 строк, а также информацию о таблице. 

In [18]:
import pandas as pd

In [19]:
data = pd.read_csv('datasets/data.csv')

In [20]:
data.head(15)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose  
0   253875.639453                      покупка жилья  
1   112080.014102            приобретение автомобиля  
2   145885.952297                      покупка жилья  
3   267628.550329         дополнительное образование  
4   158616.077870                    сыграть свадьбу  
5   255763.565419                      покупка жилья  
6   240525.971920                  операции с жильем  
7   135823.934197                        образование  
8    95856.832424              на проведение свадьбы  
9   144425.938277            покупка жилья для семьи  
10  113943.491460               покупка недвижимости  
11   77069.234271  покупка коммерческой недвижимости  
12            NaN                    сыграть свадьбу  
13  130458.228857            приобретение автомобиля  
14  165127.911772         покупка жилой недвижимости

In [21]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [22]:
# Определяем процент пропусков по каждому столбцу
data.isna().sum() / data.shape[0] * 100

children             0.000000
days_employed       10.099884
dob_years            0.000000
education            0.000000
education_id         0.000000
family_status        0.000000
family_status_id     0.000000
gender               0.000000
income_type          0.000000
debt                 0.000000
total_income        10.099884
purpose              0.000000
dtype: float64

In [23]:
data.describe().T

count           mean            std           min  \
children          21525.0       0.538908       1.381587     -1.000000   
days_employed     19351.0   63046.497661  140827.311974 -18388.949901   
dob_years         21525.0      43.293380      12.574584      0.000000   
education_id      21525.0       0.817236       0.548138      0.000000   
family_status_id  21525.0       0.972544       1.420324      0.000000   
debt              21525.0       0.080883       0.272661      0.000000   
total_income      19351.0  167422.302208  102971.566448  20667.263793   

                            25%            50%            75%           max  
children               0.000000       0.000000       1.000000  2.000000e+01  
days_employed      -2747.423625   -1203.369529    -291.095954  4.017554e+05  
dob_years             33.000000      42.000000      53.000000  7.500000e+01  
education_id           1.000000       1.000000       1.000000  4.000000e+00  
family_status_id       0.000000       0.000000       1.000000  4.000000e+00  
debt                   0.000000       0.000000       0.000000  1.000000e+00  
total_income      103053.152913  145017.937533  203435.067663  2.265604e+06

Итак, в таблице 12 столбцов.
Типы данных в столбцах: `object`, `float64`, `int64`. 

Согласно документации к данным:
* `children` — количество детей в семье
* `days_employed` — общий трудовой стаж в днях
* `dob_years` — возраст клиента в годах
* `education` — уровень образования клиента
* `education_id` — идентификатор уровня образования
* `family_status` — семейное положение
* `family_status_id` — идентификатор семейного положения
* `gender` — пол клиента
* `income_type` — тип занятости
* `debt` — имел ли задолженность по возврату кредитов
* `total_income` — ежемесячный доход
* `purpose` — цель получения кредита

В названиях столбцов нет нарушений стиля, значит, в корректировке они не нуждаются. 

В данных о трудовом стаже и ежемесячном доходе 10% пропусков. Также в данных есть ошибки, например, в колонке с трудовым стажем есть отрицательные значения, а в колонке с уровнем образования данные указаны в разных регистрах. В дальнейшем обработаем эти ошибки. 

**Выводы**

В каждой строке таблицы — данные о клиентах, их семейное положение, количество детей, наличие задолженностей по возврату кредитов. 

Предварительно можно утверждать, что данных достаточно для достижения цели исследования. Но встречаются пропуски, а также артефакты в данных.

Чтобы двигаться дальше, нужно устранить проблемы в данных.

## Предобработка данных

###  Заполнение пропусков

Выше мы выяснили, что имеются пропуски в данных с общим трудовым стажем и ежемесячным доходом, и они составляют 10%.

Посмотрим на строки с пропусками, чтобы понять их причину.

In [24]:
data_nan = data[data['days_employed'].isnull() & data['total_income'].isnull()]
display(data_nan)

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

In [25]:
#посмотрим тип занятости в пропущенных значениях
data_nan['income_type'].unique() 

array(['пенсионер', 'госслужащий', 'компаньон', 'сотрудник',
       'предприниматель'], dtype=object)

Среди пропусков нет безработных клиентов, вероятно, пропуски связаны с тем, что клиенты получают заработную плату либо пенсию в данном банке и поэтому в анкетах не указали доход.

Заменим пропущенные значения. В случае с количественными переменными наилучшие решение для замены - метод медианы `median()`. 
Среднее значение некорректно характеризует данные, когда некоторые значения сильно выделяются среди большинства (особенно, когда рассматриваются заработные платы), поэтому метод среднего арифметического нам не подойдет. 

In [26]:
#посчитаем медианные значения для трудового стажа и дохода
days_employed_median = data['days_employed'].median()
print(days_employed_median)
total_income_median = data['total_income'].median()
print(total_income_median)

-1203.369528770489
145017.93753253992


In [27]:
#заменяем пропуски на полученные медианные значения
data['days_employed'].fillna(days_employed_median, inplace=True)
data['total_income'].fillna(total_income_median, inplace=True)

In [28]:
data.head(15)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0   -1203.369529         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose  
0   253875.639453                      покупка жилья  
1   112080.014102            приобретение автомобиля  
2   145885.952297                      покупка жилья  
3   267628.550329         дополнительное образование  
4   158616.077870                    сыграть свадьбу  
5   255763.565419                      покупка жилья  
6   240525.971920                  операции с жильем  
7   135823.934197                        образование  
8    95856.832424              на проведение свадьбы  
9   144425.938277            покупка жилья для семьи  
10  113943.491460               покупка недвижимости  
11   77069.234271  покупка коммерческой недвижимости  
12  145017.937533                    сыграть свадьбу  
13  130458.228857            приобретение автомобиля  
14  165127.911772         покупка жилой недвижимости

**Выводы**

В данных имеются пропуски в колонках с трудовым стажем и доходом, их количество совпадает. Возможно, причина пропусков в том, что клиенты, получающие пенсию либо зарплату в данном банке, не указали в анкете свой доход. Пропуски мы заменили на медианные значения.

###  Проверка данных на аномалии и исправления.

В данных встречаются артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. Например, отрицательное количество дней трудового стажа или слишком большие значения в столбце `days_employed`. Вероятно, это связано с техническим сбоем при загрузке данных.

Уберем отрицательные значения в столбце со стажем, применим к нему функцию `abs()`.

In [29]:
data['days_employed'] = abs(data['days_employed'])

Проверим, есть ли значения больше 22000. Трудовой стаж больше 22000 дней (это примерно 60 лет) вряд ли возможен, клиенты старше 75 лет не могут обращаться за кредитом.

In [30]:
data.loc[(data['days_employed'] > 22000)]

children  days_employed  dob_years education  education_id  \
4             0  340266.072047         53   среднее             1   
18            0  400281.136913         53   среднее             1   
24            1  338551.952911         57   среднее             1   
25            0  363548.489348         67   среднее             1   
30            1  335581.668515         62   среднее             1   
...         ...            ...        ...       ...           ...   
21505         0  338904.866406         53   среднее             1   
21508         0  386497.714078         62   среднее             1   
21509         0  362161.054124         59    высшее             0   
21518         0  373995.710838         59   СРЕДНЕЕ             1   
21521         0  343937.404131         67   среднее             1   

               family_status  family_status_id gender income_type  debt  \
4           гражданский брак                 1      F   пенсионер     0   
18            вдовец / вдова                 2      F   пенсионер     0   
24     Не женат / не замужем                 4      F   пенсионер     0   
25           женат / замужем                 0      M   пенсионер     0   
30           женат / замужем                 0      F   пенсионер     0   
...                      ...               ...    ...         ...   ...   
21505       гражданский брак                 1      M   пенсионер     0   
21508        женат / замужем                 0      M   пенсионер     0   
21509        женат / замужем                 0      M   пенсионер     0   
21518        женат / замужем                 0      F   пенсионер     0   
21521        женат / замужем                 0      F   пенсионер     0   

        total_income                                purpose  
4      158616.077870                        сыграть свадьбу  
18      56823.777243     на покупку подержанного автомобиля  
24     290547.235997  операции с коммерческой недвижимостью  
25      55112.757732                   покупка недвижимости  
30     171456.067993  операции с коммерческой недвижимостью  
...              ...                                    ...  
21505   75439.993167                        сыграть свадьбу  
21508   72638.590915                           недвижимость  
21509   73029.059379               операции с недвижимостью  
21518  153864.650328                   сделка с автомобилем  
21521  155999.806512                   сделка с автомобилем  

[3445 rows x 12 columns]

Значение стажа не слишком важно для исследования, аномальные значения заменим медианой. 

In [31]:
data.loc[(data['days_employed'] > 22000), 'days_employed'] = data['days_employed'].median() 

In [32]:
# Проверим, что замена прошла успешно
data.loc[(data['days_employed'] > 22000)]

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

Проверим возраст заемщиков, он не должен быть меньше 18 или больше 75 лет. 

In [33]:
data.loc[(data['dob_years'] < 18) | (data['dob_years'] > 75)] 

children  days_employed  dob_years education  education_id  \
99            0    1808.053434          0   Среднее             1   
149           0    2664.273168          0   среднее             1   
270           3    1872.663186          0   среднее             1   
578           0    1808.053434          0   среднее             1   
1040          0    1158.029561          0    высшее             0   
...         ...            ...        ...       ...           ...   
19829         0    1203.369529          0   среднее             1   
20462         0    1808.053434          0   среднее             1   
20577         0    1808.053434          0   среднее             1   
21179         2     108.967042          0    высшее             0   
21313         0    1268.487728          0   среднее             1   

               family_status  family_status_id gender income_type  debt  \
99           женат / замужем                 0      F   пенсионер     0   
149                в разводе                 3      F   сотрудник     0   
270          женат / замужем                 0      F   сотрудник     0   
578          женат / замужем                 0      F   пенсионер     0   
1040               в разводе                 3      F   компаньон     0   
...                      ...               ...    ...         ...   ...   
19829        женат / замужем                 0      F   сотрудник     0   
20462        женат / замужем                 0      F   пенсионер     0   
20577  Не женат / не замужем                 4      F   пенсионер     0   
21179        женат / замужем                 0      M   компаньон     0   
21313  Не женат / не замужем                 4      M   сотрудник     0   

        total_income                                 purpose  
99      71291.522491                              автомобиль  
149     70176.435951                       операции с жильем  
270    102166.458894                            ремонт жилью  
578     97620.687042  строительство собственной недвижимости  
1040   303994.134987                         свой автомобиль  
...              ...                                     ...  
19829  145017.937533                                   жилье  
20462  259193.920299                    покупка своего жилья  
20577  129788.762899                            недвижимость  
21179  240702.007382        строительство жилой недвижимости  
21313  152419.184851                       операции с жильем  

[101 rows x 12 columns]

In [34]:
data.loc[(data['dob_years'] < 18) | (data['dob_years'] > 75), 'dob_years'].unique() 

array([0], dtype=int64)

Мы видим, что в колонке `dob_years` есть аномальные значения, все они равны 0. Вероятно, возраст просто не указан или утерян. Этот показатель не очень важен для исследования, заменим его медианным значением.

In [35]:
data.loc[(data['dob_years'] < 18) | (data['dob_years'] > 75), 'dob_years'] = data['dob_years'].median()

In [36]:
# Проверяем, что аномалий по возрасту нет
data.loc[(data['dob_years'] < 18) | (data['dob_years'] > 75), 'dob_years'].unique() 

array([], dtype=int64)

Проверим, есть ли необычные значения среди дохода. Посмотрим значения ниже прожиточного минимума или слишком большие. 

In [37]:
data.loc[(data['total_income'] < 12000) | (data['total_income'] > 10_000_000)] 

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

В `total_income` артефактов не обнаружено. 

Теперь посмотрим значения в `children`, количество детей не может быть меньше 0, также проверим большие значения.

In [38]:
data.loc[(data['children'] < 0) | (data['children'] > 9)]

children  days_employed  dob_years education  education_id  \
291          -1    4417.703588         46   среднее             1   
606          20     880.221113         21   среднее             1   
705          -1     902.084528         50   среднее             1   
720          20     855.595512         44   среднее             1   
742          -1    3174.456205         57   среднее             1   
...         ...            ...        ...       ...           ...   
21140        -1    1422.668059         44   среднее             1   
21325        20     601.174883         37   среднее             1   
21390        20    1203.369529         53   среднее             1   
21404        20     494.788448         52   среднее             1   
21491        20     173.954460         27   среднее             1   

          family_status  family_status_id gender  income_type  debt  \
291    гражданский брак                 1      F    сотрудник     0   
606     женат / замужем                 0      M    компаньон     0   
705     женат / замужем                 0      F  госслужащий     0   
720     женат / замужем                 0      F    компаньон     0   
742     женат / замужем                 0      F    сотрудник     0   
...                 ...               ...    ...          ...   ...   
21140   женат / замужем                 0      F    компаньон     0   
21325   женат / замужем                 0      F    компаньон     0   
21390   женат / замужем                 0      M    компаньон     0   
21404   женат / замужем                 0      M    компаньон     0   
21491   женат / замужем                 0      F    сотрудник     0   

        total_income                             purpose  
291    102816.346412              профильное образование  
606    145334.865002                       покупка жилья  
705    137882.899271             приобретение автомобиля  
720    112998.738649                покупка недвижимости  
742     64268.044444          дополнительное образование  
...              ...                                 ...  
21140  169562.091999     операции со своей недвижимостью  
21325  102986.065978              профильное образование  
21390  145017.937533          покупка жилой недвижимости  
21404  156629.683642     операции со своей недвижимостью  
21491  155520.899688  на покупку подержанного автомобиля  

[123 rows x 12 columns]

In [39]:
data.loc[(data['children'] < 0) | (data['children'] > 9), 'children'].unique()

array([-1, 20], dtype=int64)

Мы видим здесь аномальные значения -1 и 20. Вероятно, это техничский сбой. 
Для исследования нам важно знать количество детей у заемщика, поэтому эти данные мы не можем заменить каким-то средним значением. Но т.к. этих данных всего 123 на 21000 (<1%), их удаление не повлияет значительно на результат исследования. 

Удаляем строки с неподходящими значениями и проверяем результат.

In [40]:
data = data.drop(data[(data['children'] < 0) | (data['children'] > 9)].index).reset_index(drop=True)

In [41]:
data.loc[(data['children'] < 0) | (data['children'] > 9), 'children'].unique()

array([], dtype=int64)

Проверим значения `gender` на аномалии.

In [42]:
data['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

In [43]:
data.loc[(data['gender'] == 'XNA')]

children  days_employed  dob_years            education  education_id  \
10646         0    2358.600502         24  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10646  гражданский брак                 1    XNA   компаньон     0   

        total_income               purpose  
10646  203905.157261  покупка недвижимости

Есть одно странное значение пола, удалим его.

In [44]:
data = data.drop(data[data['gender'] == 'XNA'].index).reset_index(drop=True)

In [45]:
data.loc[(data['gender'] == 'XNA')]

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

**Выводы**

Мы выполнили проверку данных на артефакты. Отрицательные и слишком большие значения трудового стажа, а также ошибки в возрасте заменили на медианные значения. Вероятной причиной возникновения такиих ошибок могут быть технические сбои при обработке или передаче данных.

Аномальные данные также обнаружены в колонке с количеством детей заемщика. Эти данные важны для исследования, поэтому мы не могли их заменить, но т.к. число артефактов не превышает 1%, мы удалили их без значительного ущерба для релузьтатолв исследования. Так же удаленно одно значение с аномальным значением пола.

###  Изменение типов данных.

Значения в столбце с доходом и стажем заменим на целочисленные методом `astype()`

In [46]:
data['days_employed'] = data['days_employed'].astype('int')
data['total_income'] = data['total_income'].astype('int')

In [47]:
data.head()

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   Среднее             1   
3         3           4124         32   среднее             1   
4         0           1808         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

###  Удаление дубликатов.

Проверим данные на наличие дубликатов и удалим их.

In [48]:
data.duplicated().sum()

55

In [49]:
data = data.drop_duplicates().reset_index(drop=True)

In [50]:
# Убедимся, что дубликатов больше нет
data.duplicated().sum()

0

Посмотрим, какие есть дубликаты в колонках `education`, `family_status`, `income_type`, `purpose`.

In [51]:
data['education'].sort_values().unique()

array(['ВЫСШЕЕ', 'Высшее', 'НАЧАЛЬНОЕ', 'НЕОКОНЧЕННОЕ ВЫСШЕЕ',
       'Начальное', 'Неоконченное высшее', 'СРЕДНЕЕ', 'Среднее',
       'УЧЕНАЯ СТЕПЕНЬ', 'Ученая степень', 'высшее', 'начальное',
       'неоконченное высшее', 'среднее', 'ученая степень'], dtype=object)

In [52]:
data['family_status'].sort_values().unique()

array(['Не женат / не замужем', 'в разводе', 'вдовец / вдова',
       'гражданский брак', 'женат / замужем'], dtype=object)

In [53]:
data['income_type'].sort_values().unique()

array(['безработный', 'в декрете', 'госслужащий', 'компаньон',
       'пенсионер', 'предприниматель', 'сотрудник', 'студент'],
      dtype=object)

In [54]:
data['purpose'].sort_values().unique()

array(['автомобили', 'автомобиль', 'высшее образование',
       'дополнительное образование', 'жилье',
       'заняться высшим образованием', 'заняться образованием',
       'на покупку автомобиля', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля', 'на проведение свадьбы',
       'недвижимость', 'образование', 'операции с жильем',
       'операции с коммерческой недвижимостью',
       'операции с недвижимостью', 'операции со своей недвижимостью',
       'покупка жилой недвижимости', 'покупка жилья',
       'покупка жилья для сдачи', 'покупка жилья для семьи',
       'покупка коммерческой недвижимости', 'покупка недвижимости',
       'покупка своего жилья', 'получение высшего образования',
       'получение дополнительного образования', 'получение образования',
       'приобретение автомобиля', 'профильное образование',
       'ремонт жилью', 'свадьба', 'свой автомобиль',
       'сделка с автомобилем', 'сделка с подержанным автомобилем',
       'строительство 

В столбцах `family_status` и `income_type` нет дубликатов. 

В колонке `purpose` нет дубликатов, связанных с написанием в разных регистрах, но есть схожие по смыслу цели. В дальнейшем объединим их в категории.

В столбце `education` есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Это неявные дубликаты. Их появление связано, скорее всего, с человеческим фактором. Приведем их к одному регистру. 

In [55]:
data['education'] = data['education'].str.lower() 

In [56]:
# Убедимся в отсутствии неявных дубликатов
data['education'].sort_values().unique()

array(['высшее', 'начальное', 'неоконченное высшее', 'среднее',
       'ученая степень'], dtype=object)

Мы удалили явные дубликаты методом `drop_duplicates()`. 
Также мы удалили неявные дубликаты, связанные с написанием в разных регистрах в колонке `education`, приведением их к одному регистру методом `lower()`.

###  Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

Создадим два новых датафрейма со столбцами:
`education_id` и `education` — в первом;
`family_status_id` и `family_status` — во втором. 
Они будут "словарями", к которым мы сможем обращаться с помощью индектификатора. 

Из исходного датафрейма удалим столбцы `education` и `family_status`, оставив только их идентификаторы: `education_id` и `family_status_id`, таким образом сделаем нашу таблицу компактнее и удобней для работы. 

Создаем таблицу-словарь с уровнем образования `education_data`, удаляем из нее дубликаты. 

In [57]:
education_data = data[['education_id', 'education']]
education_data.head()

education_id education
0             0    высшее
1             1   среднее
2             1   среднее
3             1   среднее
4             1   среднее

In [58]:
education_data.duplicated().sum()

21341

In [59]:
education_data = education_data.drop_duplicates().reset_index(drop=True)

In [60]:
education_data.duplicated().sum()

0

In [61]:
# Посмотрим на полученную таблицу
education_data

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

Создаем таблицу-словарь с семейным положением `family_status_data`, удаляем все дубликаты и отображаем результат.

In [62]:
family_status_data = data[['family_status_id', 'family_status']]
family_status_data.head()

family_status_id     family_status
0                 0   женат / замужем
1                 0   женат / замужем
2                 0   женат / замужем
3                 0   женат / замужем
4                 1  гражданский брак

In [63]:
family_status_data.duplicated().sum()

21341

In [64]:
family_status_data = family_status_data.drop_duplicates().reset_index(drop=True)

In [65]:
family_status_data

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  Не женат / не замужем

Теперь из исходной таблицы `data` удаляем лишние столбцы, которые перенесли в словари: `education` и `family_status`

In [66]:
data = data.drop(['education', 'family_status'], axis=1)
data.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0           1808         53             1                 1      F   

  income_type  debt  total_income                     purpose  
0   сотрудник     0        253875               покупка жилья  
1   сотрудник     0        112080     приобретение автомобиля  
2   сотрудник     0        145885               покупка жилья  
3   сотрудник     0        267628  дополнительное образование  
4   пенсионер     0        158616             сыграть свадьбу

###  Категоризация дохода.

У нас большое количество разных вариантов дохода. 
Работать с единичными отрывками и делать из них статистические выводы нельзя. Значит, нужно сгруппировать данные, чтобы численности каждой группы хватало для формулировки выводов.
Нужна категоризация — объединение данных в категории. 

Создадим столбец `total_income_category` с категоризацией доходов:
     
* 0–30000 — `'E'`;
* 30001–50000 — `'D'`;
* 50001–200000 — `'C'`;
* 200001–1000000 — `'B'`;
* 1000001 и выше — `'A'`

In [67]:
# Для начала напишем функцию, которая разделит значения доходов по категориям
def income_category(income):
    if  income <= 30000:
        return 'E'
    if income <= 50000:
        return 'D'
    if income <= 200_000:
        return 'C'
    if income <= 1000_000:
        return 'B'
    return 'A'
  
print(income_category(200000))
print(income_category(1000001))

C
A


In [68]:
# Теперь создадим новый столбец, применяя к ['total_income'] функцию income_category
data['total_income_category'] = data['total_income'].apply(income_category)
data.head(10)

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0           1808         53             1                 1      F   
5         0            926         27             0                 1      M   
6         0           2879         43             0                 0      F   
7         0            152         50             1                 0      M   
8         2           6929         35             0                 1      F   
9         0           2188         41             1                 0      M   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   
5   компаньон     0        255763               покупка жилья   
6   компаньон     0        240525           операции с жильем   
7   сотрудник     0        135823                 образование   
8   сотрудник     0         95856       на проведение свадьбы   
9   сотрудник     0        144425     покупка жилья для семьи   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C  
5                     B  
6                     B  
7                     C  
8                     C  
9                     C

In [69]:
data['total_income_category'].value_counts()

C    15938
B     5012
D      349
A       25
E       22
Name: total_income_category, dtype: int64

Наиболее распространенная категория доходов среди клиентов банка  - `C` от 50 до 200 тыс. 

###  Категоризация целей кредита.

Теперь объединим различные цели получения кредита по категориям.
Создадим функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:
* `'операции с автомобилем'`,
* `'операции с недвижимостью'`,
* `'проведение свадьбы'`,
* `'получение образования`'

In [70]:
# Посмотрим, какие есть цели кредита в наших исходных данных
data['purpose'].sort_values().unique()

array(['автомобили', 'автомобиль', 'высшее образование',
       'дополнительное образование', 'жилье',
       'заняться высшим образованием', 'заняться образованием',
       'на покупку автомобиля', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля', 'на проведение свадьбы',
       'недвижимость', 'образование', 'операции с жильем',
       'операции с коммерческой недвижимостью',
       'операции с недвижимостью', 'операции со своей недвижимостью',
       'покупка жилой недвижимости', 'покупка жилья',
       'покупка жилья для сдачи', 'покупка жилья для семьи',
       'покупка коммерческой недвижимости', 'покупка недвижимости',
       'покупка своего жилья', 'получение высшего образования',
       'получение дополнительного образования', 'получение образования',
       'приобретение автомобиля', 'профильное образование',
       'ремонт жилью', 'свадьба', 'свой автомобиль',
       'сделка с автомобилем', 'сделка с подержанным автомобилем',
       'строительство 

In [71]:
def purpose_category(purpose):
    if 'авто' in purpose:
        return 'операции с автомобилем'
    if 'жиль' in purpose or 'недвиж' in purpose:
        return 'операции с недвижимостью'
    if 'свадьб' in purpose:
        return 'проведение свадьбы'
    if 'образов' in purpose: 
        return 'получение образования'
    return 'цель неопределена'
    
    
print(purpose_category('сыграть свадьбу'))    
print(purpose_category('на покупку подержанного автомобиля'))
print(purpose_category('на покупку жилья'))
print(purpose_category('строительство недвижимости'))
print(purpose_category('сыграть свадьбу'))
print(purpose_category('получение дополнительного образования'))
print(purpose_category('купить лотерейные билеты'))
    

проведение свадьбы
операции с автомобилем
операции с недвижимостью
операции с недвижимостью
проведение свадьбы
получение образования
цель неопределена


In [72]:
data['purpose_category'] = data['purpose'].apply(purpose_category)
data.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0           1808         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category          purpose_category  
0                     B  операции с недвижимостью  
1                     C    операции с автомобилем  
2                     C  операции с недвижимостью  
3                     B     получение образования  
4                     C        проведение свадьбы

In [73]:
data['purpose_category'].value_counts()

операции с недвижимостью    10753
операции с автомобилем       4281
получение образования        3989
проведение свадьбы           2323
Name: purpose_category, dtype: int64

Среди целей кредита лидируют операции с недвижимостью. 

## Исследовательский анализ. Ответы на вопросы.

### Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?

Ответим на вопросы исследования

* Есть ли зависимость между количеством детей и возвратом кредита в срок?

* Есть ли зависимость между семейным положением и возвратом кредита в срок?

* Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

* Как разные цели кредита влияют на его возврат в срок?

* Есть ли зависимость между уровнем образования и возвратом кредита в срок?

* Есть ли зависимость между типом занятости и возвратом кредита в срок?

Выясним, есть ли зависимость между количеством детей и возвратом кредита в срок. Для этого создадим сводную таблицу `data_pivot_children`, в которой отобразим количество вовремя погашенных кредитов `debt 0` и количество просрочек `debt 1` в зависимости от количества детей `children`

In [74]:
data_pivot_children = data.pivot_table(index=['children'], columns='debt', values='purpose', aggfunc='count')

In [75]:
data_pivot_children

debt            0       1
children                 
0         13042.0  1063.0
1          4365.0   444.0
2          1858.0   194.0
3           303.0    27.0
4            37.0     4.0
5             9.0     NaN

In [76]:
# Заменим отсуствующие значения на 0
data_pivot_children.fillna(0, inplace=True)

Для наглядности добавим столбец `ratio`, в котором посчитаем процент своевременных возвратов от общего числа кредитов.

In [77]:
data_pivot_children['ratio'] = data_pivot_children[0] / (data_pivot_children[0] + data_pivot_children[1])*100
data_pivot_children

debt            0       1       ratio
children                             
0         13042.0  1063.0   92.463665
1          4365.0   444.0   90.767311
2          1858.0   194.0   90.545809
3           303.0    27.0   91.818182
4            37.0     4.0   90.243902
5             9.0     0.0  100.000000

В этой таблице мы видим, что процент возратов у клиентов без детей немного выше, чем у клиентов с 3 детьми. У клиентов с 1, 2 и 4 детьми примерно своевременный возврат примерно на 2% ниже, чем у бездетных клиентов. У клиентов с 5 детьми 100% возврат, но их количество слишком мало, чтобы использовать в выводах. 

#### Вывод 1:

Исходя из вышесказанного можно сделать вывод, что клиенты без детей или с 3 детьми на 2% привлекательней для банка в качестве заемщиков. 

### Вопрос 2:Есть ли зависимость между семейным положением и возвратом кредита в срок?

Выясним, есть ли зависимость между семейным положением и возвратом кредита в срок. Для этого создадим сводную таблицу `data_pivot_family_id`, в которой отобразим количество вовремя погашенных кредитов `debt 0` и количество просрочек `debt 1` в зависимости от семейного положения `family_id`.

Создадим временную таблицу `data_final_f`, добавив данные с расшифровкой семейного статуса для наглядности. 

In [78]:
data_final_f = data.merge(family_status_data, on='family_status_id', how='right')

In [79]:
data_final_f.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0           2879         43             0                 0      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   компаньон     0        240525           операции с жильем   

  total_income_category          purpose_category    family_status  
0                     B  операции с недвижимостью  женат / замужем  
1                     C    операции с автомобилем  женат / замужем  
2                     C  операции с недвижимостью  женат / замужем  
3                     B     получение образования  женат / замужем  
4                     B  операции с недвижимостью  женат / замужем

In [80]:
# Формируем сводную таблицу
data_pivot_family_status = data_final_f.pivot_table(
    index=['family_status_id', 'family_status'], 
    columns='debt', 
    values='purpose', 
    aggfunc='count'
)

In [81]:
data_pivot_family_status

debt                                        0    1
family_status_id family_status                    
0                женат / замужем        11339  927
1                гражданский брак        3759  385
2                вдовец / вдова           888   63
3                в разводе               1105   84
4                Не женат / не замужем   2523  273

In [82]:
# Добавляем столбец с процентом своевременных возвратов для наглядности
data_pivot_family_status['ratio'] = data_pivot_family_status[0] / (
    data_pivot_family_status[0] + data_pivot_family_status[1])*100

In [83]:
data_pivot_family_status

debt                                        0    1      ratio
family_status_id family_status                               
0                женат / замужем        11339  927  92.442524
1                гражданский брак        3759  385  90.709459
2                вдовец / вдова           888   63  93.375394
3                в разводе               1105   84  92.935240
4                Не женат / не замужем   2523  273  90.236052

Здесь мы видим, что наиболее высокий процент возвратов у вдовцов (но их количество также значительно меньше остальных категорий семейного положения). У состоящих в браке и разведенных процент возврата на 2% выше, чем у не состоящих в браке либо состоящих в гражданском браке. 

#### Вывод 2:

Для банка в качестве заемщиков более привлекательны (на 2%) клиенты, состоящие в браке, в разводе и вдовцы. 

### Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Ответим на вопрос 
* Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Составим сводную таблицу `data_pivot_income` с зависимостью возврата кредита в срок от уровня дохода. 

In [84]:
data_pivot_income = data.pivot_table(index=['total_income_category'], columns='debt', values='purpose', aggfunc='count')

In [85]:
data_pivot_income

debt                       0     1
total_income_category             
A                         23     2
B                       4658   354
C                      14585  1353
D                        328    21
E                         20     2

In [86]:
# Как и ранее, добавляем столбец с процентным отношением возвратов в срок к общему числу кредитов
data_pivot_income['ratio'] = data_pivot_income[0] / (data_pivot_income[0] + data_pivot_income[1])*100

In [87]:
data_pivot_income

debt                       0     1      ratio
total_income_category                        
A                         23     2  92.000000
B                       4658   354  92.936951
C                      14585  1353  91.510855
D                        328    21  93.982808
E                         20     2  90.909091

Здесь мы видим, что наиболее высокий процент вовзрата у клиентов с доходом `D` (30001–50000) и  `B` (200001–1000000). Клиенты с уровнем дохода `A`(1000001 и выше),  `C` (50001-200000) и `E` (0–30000) примерно на 2% реже возвращают кредиты в срок.

#### Вывод 3:

Для банка наиболее привлекательны в качестве заемщиков клиенты с доходом 30-50 тыс. и 200-1000 тыс. 

### Вопрос 4: Как разные цели кредита влияют на его возврат в срок?

Ответим на вопрос 
* Как разные цели кредита влияют на его возврат в срок?

Создаем сводную таблицу `data_pivot_purpose` с зависимостью возврата кредитов в срок от цели кредита.

In [88]:
data_pivot_purpose = data.pivot_table(index=['purpose_category'], columns='debt', values='purpose', aggfunc='count')

In [89]:
data_pivot_purpose

debt                         0    1
purpose_category                   
операции с автомобилем    3881  400
операции с недвижимостью  9973  780
получение образования     3620  369
проведение свадьбы        2140  183

Добавляем столбец `ratio` с процентами своевременных возвратов кредитов.

In [90]:
data_pivot_purpose['ratio'] = data_pivot_purpose[0] / (data_pivot_purpose[0] + data_pivot_purpose[1])*100

In [91]:
data_pivot_purpose

debt                         0    1      ratio
purpose_category                              
операции с автомобилем    3881  400  90.656389
операции с недвижимостью  9973  780  92.746210
получение образования     3620  369  90.749561
проведение свадьбы        2140  183  92.122256

Из таблицы мы видим, что кредиты, взятые для операций с недвижимостью или на проведение свадьбы возвращаются на 2% чаще, чем кредиты на операции с автомобилем или получение образования.

#### Вывод 4:

Кредиты, полученные на операции с недвижимостью и проведение свадьбы, возвращают в срок на 2% чаще, чем кредиты на другие цели.

### Вопрос 5: Есть ли зависимость между уровнем образования и возвратом кредита в срок?

Ответим на вопрос 
* Есть ли зависимость между уровнем образования и возвратом кредита в срок?

Создадим временную таблицу `data_final_e`, добавив к нашей таблице расшифровку категорий образования для наглядности.

In [92]:
data_final_e = data.merge(education_data, on='education_id', how='right')

In [93]:
data_final_e.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         0            926         27             0                 1      M   
2         0           2879         43             0                 0      F   
3         2           6929         35             0                 1      F   
4         2           4171         36             0                 0      M   

  income_type  debt  total_income                purpose  \
0   сотрудник     0        253875          покупка жилья   
1   компаньон     0        255763          покупка жилья   
2   компаньон     0        240525      операции с жильем   
3   сотрудник     0         95856  на проведение свадьбы   
4   компаньон     0        113943   покупка недвижимости   

  total_income_category          purpose_category education  
0                     B  операции с недвижимостью    высшее  
1                     B  операции с недвижимостью    высшее  
2                     B  операции с недвижимостью    высшее  
3                     C        проведение свадьбы    высшее  
4                     C  операции с недвижимостью    высшее

In [94]:
# Формираем сводную таблицу
data_pivot_education = data_final_e.pivot_table(
    index=['education_id', 'education'], 
    columns='debt', 
    values='purpose', 
    aggfunc='count'
)

In [95]:
data_pivot_education

debt                                    0       1
education_id education                           
0            высшее                4950.0   278.0
1            среднее              13735.0  1355.0
2            неоконченное высшее    672.0    68.0
3            начальное              251.0    31.0
4            ученая степень           6.0     NaN

In [96]:
# Убираем NaN
data_pivot_education.fillna(0, inplace=True)

In [97]:
data_pivot_education

debt                                    0       1
education_id education                           
0            высшее                4950.0   278.0
1            среднее              13735.0  1355.0
2            неоконченное высшее    672.0    68.0
3            начальное              251.0    31.0
4            ученая степень           6.0     0.0

In [98]:
# Считаем проценты вовзрата и вносим в таблицу
data_pivot_education['ratio'] = data_pivot_education[0] / (data_pivot_education[0] + data_pivot_education[1])*100

In [99]:
data_pivot_education

debt                                    0       1       ratio
education_id education                                       
0            высшее                4950.0   278.0   94.682479
1            среднее              13735.0  1355.0   91.020543
2            неоконченное высшее    672.0    68.0   90.810811
3            начальное              251.0    31.0   89.007092
4            ученая степень           6.0     0.0  100.000000

#### Вывод 5:

Здесь явно прослеживается зависимость: чем выше уровень образования, тем выше процент своевременных возвратов по кредиту. 

### Вопрос 6: Есть ли зависимость между типом занятости и возвратом кредита в срок?

Ответим на вопрос
* Есть ли зависимость между типом занятости и возвратом кредита в срок?

In [100]:
# Создаем сводную таблицу
data_pivot_income_type = data.pivot_table(index=['income_type'], columns='debt', values='purpose', aggfunc='count')

In [101]:
data_pivot_income_type

debt                  0       1
income_type                    
безработный         1.0     1.0
в декрете           NaN     1.0
госслужащий      1365.0    86.0
компаньон        4674.0   374.0
пенсионер        3604.0   216.0
предприниматель     2.0     NaN
сотрудник        9967.0  1054.0
студент             1.0     NaN

In [102]:
# Убираем NaN
data_pivot_income_type.fillna(0, inplace=True)

In [103]:
data_pivot_income_type

debt                  0       1
income_type                    
безработный         1.0     1.0
в декрете           0.0     1.0
госслужащий      1365.0    86.0
компаньон        4674.0   374.0
пенсионер        3604.0   216.0
предприниматель     2.0     0.0
сотрудник        9967.0  1054.0
студент             1.0     0.0

In [104]:
# Добавляем в сводную таблицу столбец с процентами вовзратов кредитов в срок к общему числу кредитов
data_pivot_income_type['ratio'] = data_pivot_income_type[0] / (data_pivot_income_type[0] + data_pivot_income_type[1])*100

In [105]:
data_pivot_income_type

debt                  0       1       ratio
income_type                                
безработный         1.0     1.0   50.000000
в декрете           0.0     1.0    0.000000
госслужащий      1365.0    86.0   94.073053
компаньон        4674.0   374.0   92.591125
пенсионер        3604.0   216.0   94.345550
предприниматель     2.0     0.0  100.000000
сотрудник        9967.0  1054.0   90.436440
студент             1.0     0.0  100.000000

#### Вывод 6:

Здесь можно сделать вывод, что госслужащие и пенсионеры возвращают кредиты в срок на 2% чаще, чем компаньоны и на 4% чаще, чем сотрудники. Безработных клиентов, клиентов в декрете, предпринимателей и студентов не будем учитывать из-за малого количества в выборке. 

## Общий вывод:

Мы провели исследование с целью разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. 
Так же мы оценили зависимости своевременного вовзрата кредита от других данных, предоставленных банком, которые будут учтены при построении модели кредитного скоринга.


  
    На этапе предобработки данных нами было выявлено следующее:
    
        * Были заменены пропущенные данные в трудовом стаже и ежемесячном доходе на медианы.
        * Данные были проверены на наличие аномалий, заменены и удалены аномалии в трудовом стаже, количестве детей и поле. 
        * Типы данных исправлены на корректные.
        * Удалены явные дубликаты в данных, а так же выявлены неявные дубликаты, связанные с использованием различных регистров при заполнении данных.
       
    
    На этапе расчетов \ категоризации нами были добавлены следующие данные:
        * Доход был распределен на категории.
        * Все цели кредитов были распределены на 4 категории.
    
    Далее был проведен исследовательский анализ данных и получены ответы о том, какие факторы влияют на возврат кредитов: 
       * Клиенты без детей или с 3 детьми на 2% привлекательней для банка в качестве заемщиков.
       * Клиенты, состоящие в браке, в разводе и вдовцы на 2% более привлекательныв качестве заемщиков.
       * Клиенты с доходом 30-50 тыс. и 200-1000 тыс. привлекательнее для банка в качестве заемщиков.
       * Кредиты, полученные на операции с недвижимостью и проведение свадьбы, возвращают в срок на 2% чаще, чем кредиты на другие цели.
       * Чем выше уровень образования клиента, тем выше процент своевременных возвратов по кредиту.
       * Госслужащие и пенсионеры возвращают кредиты в срок на 2% чаще, чем компаньоны и на 4% чаще, чем сотрудники.
       
    Исходя из этого можно сказать, что идеальный заемщик для банка - бездетный вдовец с доходом от 30 до 50 тыс., имеющий высшее образование или ученую степень, пенсионер, желающий совершить операцию с недвижимостью. Наименее привлекательный заемщик - холостяк с 2 детьми, доходом ниже 30 тыс. (или более 1000_000), сотрудник с начальным образованием, берущий кредит для операций с автомобилем.
    